src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.GroupedAnalyzerActions.cs

```diff
@@ -74,7 +85,35 @@ IGroupedAnalyzerActions IGroupedAnalyzerActions.Append(IGroupedAnalyzerActions i
 
                 var newGroupedActions = GroupedActionsByAnalyzer.AddRange(groupedAnalyzerActions.GroupedActionsByAnalyzer);
                 var newAnalyzerActions = AnalyzerActions.Append(groupedAnalyzerActions.AnalyzerActions);
-                return new GroupedAnalyzerActions(newGroupedActions, newAnalyzerActions);
+                var analyzersByKind = CreateAnalyzersByKind(newGroupedActions);
+                return new GroupedAnalyzerActions(newGroupedActions, analyzersByKind, newAnalyzerActions);
+            }
+
+            private static ImmutableDictionary<TLanguageKindEnum, ImmutableArray<DiagnosticAnalyzer>> CreateAnalyzersByKind(ImmutableArray<(DiagnosticAnalyzer, GroupedAnalyzerActionsForAnalyzer)> groupedActionsAndAnalyzers)
+            {
+                var analyzersByKind = PooledDictionary<TLanguageKindEnum, ArrayBuilder<DiagnosticAnalyzer>>.GetInstance();
+                foreach (var (analyzer, groupedActionsForAnalyzer) in groupedActionsAndAnalyzers)
+                {
+                    foreach (var (kind, _) in groupedActionsForAnalyzer.NodeActionsByAnalyzerAndKind)
+                    {
+                        if (!analyzersByKind.TryGetValue(kind, out var builder))
+                        {
+                            builder = ArrayBuilder<DiagnosticAnalyzer>.GetInstance();
+                            analyzersByKind.Add(kind, builder);
+                        }
+                        builder.Add(analyzer);
+                    }
+                }
+
+                // Essentially an inline of PooledBuilderExtensions.ToImmutableMultiDictionaryAndFree
+                var result = ImmutableDictionary.CreateBuilder<TLanguageKindEnum, ImmutableArray<DiagnosticAnalyzer>>();
+                foreach (var (key, items) in analyzersByKind)
+                {
+                    result.Add(key, items.ToImmutableAndFree());
+                }
+
+                analyzersByKind.Free();
+                return result.ToImmutable();
```

Comment:
i could have sworn there was a helper that alreaday did the conversion (with all the value freeing) for this sort of Dict<K, arrayBuilder> type.

Reply:
I saw IDictionaryExtensions.MultiAdd and PooledBuilderExtensions.ToImmutableMultiDictionaryAndFree that I mentioned in the comments, but I couldn't get to those from the compiler. I definitely could have missed something, but I did look around.

Reply:
ah.  just move to a common shared code location.  we have tons.  it will require compiler signoff.  but given the niceness of the pattern, i think that should be fine.

Reply:
Found something. AddPooled and ToImmutableSegmentedDictionaryAndFree both exist under the compilers folder, I'm fine switching this to ImmutableSegmentedDictionary